### PR TITLE
ARROW-4687: [Python] Stop Flight server on incoming signals

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by lihalite** [comment]
Would it be possible to add this patch so that the integration tests still run? (They wait for a certain string to be printed to detect server startup.) Otherwise, this looks good to me.

```patch
From d171f92652767b76b48636d4a2fda215347b9422 Mon Sep 17 00:00:00 2001
From: David Li <David.M.Li@twosigma.com>
Date: Wed, 27 Feb 2019 09:11:28 -0500
Subject: [PATCH] Print sentinel value in Flight integration test server

---
 cpp/src/arrow/flight/test-integration-server.cc | 1 +
 1 file changed, 1 insertion(+)

diff --git a/cpp/src/arrow/flight/test-integration-server.cc b/cpp/src/arrow/flight/test-integration-server.cc
index bc8b5cc1..c8b776e5 100644
--- a/cpp/src/arrow/flight/test-integration-server.cc
+++ b/cpp/src/arrow/flight/test-integration-server.cc
@@ -125,6 +125,7 @@ int main(int argc, char** argv) {
   ARROW_CHECK_OK(g_server->Init(FLAGS_port));
   // Exit with a clean error code (0) on SIGTERM
   ARROW_CHECK_OK(g_server->SetShutdownOnSignals({SIGTERM}));
+  std::cout << "Server listening on localhost:" << FLAGS_port << std::endl;
   ARROW_CHECK_OK(g_server->Serve());
   return 0;
 }
-- 
2.14.3 (Apple Git-98)
```

**Comment by pitrou** [comment]
@lihalite Yes, of course.

**Review by ghost — APPROVED** [review]

File: cpp/src/arrow/flight/server.cc [hunk 1]
```diff
@@ -294,30 +295,97 @@ class FlightServiceImpl : public FlightService::Service {
...
-  std::unique_ptr<grpc::Server> server;
+struct FlightServerBase::Impl {
+  std::string address_;
+  std::unique_ptr<FlightServiceImpl> service_;
+  std::unique_ptr<grpc::Server> server_;
+
+  // Signal handling
+  std::vector<int> signals_;
+  std::vector<struct sigaction> old_signal_handlers_;
+  volatile int got_signal_;
```
Comment (fsaintjacques) [inline cpp/src/arrow/flight/server.cc]: Why not atomic instead of volatile, that would make `GotSignal` concurrency safe.

Comment (pitrou) [reply cpp/src/arrow/flight/server.cc]: I don't think other methods in this class are thread-safe... That said, I can indeed use an atomic.

**Comment by pitrou** [comment]
Rebased.

**Comment by pitrou** [comment]
Will merge if no more comments.

**Pull request closed without merging**